engine/uk.co.spudsoft.birt.emitters.excel/src/uk/co/spudsoft/birt/emitters/excel/HandlerState.java

```diff
@@ -289,20 +290,83 @@ public int findRowsSpanned(int rowX, int colX) {
 		return 0;
 	}
 
+	/**
+	 * Maximum number of characters that Excel will accept for a sheet name.
+	 */
+	public static final int MAX_SHEET_NAME_LENGTH = 31;
+	/**
+	 * Characters that Excel will not accept in a sheet name.
+	 */
+	public static final String[] ILLEGAL_SHEET_NAME_CHARACTERS = new String[] { "\\", //$NON-NLS-1$
+			"/", //$NON-NLS-1$
+			"*", //$NON-NLS-1$
+			"[", //$NON-NLS-1$
+			"]", //$NON-NLS-1$
+			":", //$NON-NLS-1$
+			"?" //$NON-NLS-1$
+	};
+
+	/**
+	 * Remove illegal characters from the sheet name and make sure it's not too
+	 * long.
+	 *
+	 * @param sheetName
+	 * @return corrected sheet name
+	 */
 	public String correctSheetName(final String sheetName) {
 		// https://www.accountingweb.com/technology/excel/seven-characters-you-cant-use-in-worksheet-names
 		if (sheetName == null) {
 			return null;
 		}
-		String correctedSheetName = sheetName.replace("\\", "").replace("/", "").replace("*", "").replace("[", "")
-				.replace("]", "").replace(":", "").replace("?", "").trim();
-		if ("history".equalsIgnoreCase(correctedSheetName)) {
-			return "history-sheet";
+		String correctedSheetName = sheetName;
+		for (String illegalChar : ILLEGAL_SHEET_NAME_CHARACTERS) {
+			correctedSheetName = correctedSheetName.replace(illegalChar, ""); //$NON-NLS-1$
+		}
+		if ("history".equalsIgnoreCase(correctedSheetName)) { //$NON-NLS-1$
+			return "history-sheet"; //$NON-NLS-1$
 		}
-		// max length is 31 but leave room for digits
-		if (correctedSheetName.length() > 26) {
-			correctedSheetName = correctedSheetName.substring(0, 26);
+		if (correctedSheetName.length() > MAX_SHEET_NAME_LENGTH) {
+			correctedSheetName = correctedSheetName.substring(0, MAX_SHEET_NAME_LENGTH);
 		}
 		return correctedSheetName;
 	}
+
+	/**
+	 * Add an index to the end of the sheet name if necessary. Shorten the sheet
+	 * name if necessary. The sheet name is assumed to not be too long to start
+	 * with.
+	 *
+	 * @return the prepared sheet name
+	 * @throws BirtException
+	 */
+	public String prepareSheetName() throws BirtException {
+		String sheetName = this.sheetName;
+		if (sheetName == null) {
+			return null;
+		}
+		String preparedName = sheetName;
+		Integer previousNameCount = 1;
+		while (true) {
+			Integer nameCount = this.sheetNames.get(sheetName);
+			if (nameCount == null) {
+				this.sheetNames.put(sheetName, previousNameCount);
+				return preparedName;
+			}
+			++nameCount;
+			preparedName = sheetName + " " + nameCount; //$NON-NLS-1$
+			int correction = preparedName.length() - MAX_SHEET_NAME_LENGTH;
+			if (correction <= 0) {
+				this.sheetNames.put(sheetName, nameCount);
+				return preparedName;
+			}
+			if (correction > sheetName.length()) {
+				throw new BirtException(EmitterServices.getPluginName(),
+						"Unable to fit sheet name into the maximum allowed length", //$NON-NLS-1$
+						null);
+			}
+			sheetName = sheetName.substring(0, sheetName.length() - correction);

```

Comment:
My gut feeling says there must be a bug somewhere in this code.
The attribute sheetName is modified here, but nameCount is unchanged. Shouldn't nameCount be re-initialized then?
Not sure if I grep the code, my brains seems to go into stand-by after a hard working day...

Reply:
I thought about that.  At first I reset the name count to zero but then I thought that it might be valuable to have some indication that the tab is an increment of the previous tab even though the name is shorter.  I had considered adding ellipses to make that more clear but realized that might break existing reports, so I settled on just letting the index increment.  Make sense?

Reply:
Regarding web.xml, it looks like that file is generated.  I pushed it in hopes it had something to do with the failures but no luck.